File: app/src/main/java/ru/e6atb/chat/DisplayText.java
package ru.e6atb.chat;

import android.os.Build;

final class DisplayText {
	private DisplayText() {
	}

	static String safe(String value) {
		if (value == null || value.length() == 0) return "";
		boolean stripSupplementary = legacyEmojiLayoutBug();
		boolean changed = false;
		for (int i = 0; i < value.length(); i++) {
			char c = value.charAt(i);
			if (Character.isHighSurrogate(c)) {
				if (i + 1 < value.length() && Character.isLowSurrogate(value.charAt(i + 1))) {
					if (stripSupplementary) {
						changed = true;
						break;
					}
					i++;
				} else {
					changed = true;
					break;
				}
			} else if (Character.isLowSurrogate(c)) {
				changed = true;
				break;
			}
		}
		if (!changed) return value;
		StringBuilder out = new StringBuilder(value.length());
		for (int i = 0; i < value.length(); i++) {
			char c = value.charAt(i);
			if (Character.isHighSurrogate(c)) {
				if (i + 1 < value.length() && Character.isLowSurrogate(value.charAt(i + 1))) {
					if (stripSupplementary) {
						out.append('\uFFFD');
					} else {
						out.append(c);
						out.append(value.charAt(i + 1));
					}
					i++;
				} else {
					out.append('\uFFFD');
				}
			} else if (Character.isLowSurrogate(c)) {
				out.append('\uFFFD');
			} else {
				out.append(c);
			}
		}
		return out.toString();
	}

	private static boolean legacyEmojiLayoutBug() {
		return Build.VERSION.SDK_INT > 0 && Build.VERSION.SDK_INT < 21;
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/DisplayText.kt
package ru.e6atb.chat

import android.os.Build

object DisplayText {
    @JvmStatic
    fun safe(value: String?): String {
        if (value.isNullOrEmpty()) return ""
        val stripSupplementary = Build.VERSION.SDK_INT in 1..20
        var changed = false
        var index = 0
        while (index < value.length) {
            val character = value[index]
            when {
                Character.isHighSurrogate(character) && index + 1 < value.length && Character.isLowSurrogate(value[index + 1]) -> {
                    if (stripSupplementary) { changed = true; break }
                    index++
                }
                Character.isHighSurrogate(character) || Character.isLowSurrogate(character) -> { changed = true; break }
            }
            index++
        }
        if (!changed) return value
        return buildString(value.length) {
            var cursor = 0
            while (cursor < value.length) {
                val character = value[cursor]
                when {
                    Character.isHighSurrogate(character) && cursor + 1 < value.length && Character.isLowSurrogate(value[cursor + 1]) -> {
                        if (stripSupplementary) append('\uFFFD') else append(character).append(value[cursor + 1])
                        cursor++
                    }
                    Character.isHighSurrogate(character) || Character.isLowSurrogate(character) -> append('\uFFFD')
                    else -> append(character)
                }
                cursor++
            }
        }
    }
}
